server/src/test/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsServiceTests.java

```diff
@@ -586,6 +593,115 @@ public void testRedForNoMasterAndWithMasterEligibleNodesAndNoLeader() throws IOE
+                        .clusterFormationState();
+                    assertThat(clusterFormationState.getDescription(), not(emptyOrNullString()));
+                });
+
+                /*
+                 * Now we disconnect a random node, simulate running the cluster for a little while, and make sure that the results of
+                 * beginPollingClusterFormationInfo() contain the expected exceptions.
+                 */
+                Cluster.ClusterNode nodeToDisconnect = cluster.clusterNodes.stream()
+                    .filter(clusterNode -> clusterNode.getLocalNode().isMasterNode())
+                    .findAny()
+                    .get();
+                nodeToDisconnect.disconnect();
+                cluster.stabilise();
+                assertThat(nodeToClusterFormationStateMap.size(), equalTo(masterNodes.size()));
+                AtomicInteger exceptions = new AtomicInteger();
+                masterNodes.stream().filter(masterNode -> node.getLocalNode().equals(masterNode) == false).forEach(masterNode -> {
+                    ClusterFormationStateOrException clusterFormationStateOrException = nodeToClusterFormationStateMap.get(masterNode);
+                    assertNotNull(clusterFormationStateOrException);
+                    if (clusterFormationStateOrException.clusterFormationState() != null) {
+                        assertNull(clusterFormationStateOrException.exception());
+                        ClusterFormationFailureHelper.ClusterFormationState clusterFormationState = clusterFormationStateOrException
+                            .clusterFormationState();
+                        assertThat(clusterFormationState.getDescription(), not(emptyOrNullString()));
+                    } else {
+                        assertNotNull(clusterFormationStateOrException.exception());
+                        exceptions.getAndIncrement();
+                    }
+                });
+                if (node.equals(nodeToDisconnect)) {
+                    // If this was the disconnected node, it will have encountered exceptions contacting all nodes except itself:
+                    assertThat(exceptions.get(), equalTo(masterNodes.size() - 1));
+                } else {
+                    // Other nodes will only have encountered an exception contacting the disconnected node:
+                    assertThat(exceptions.get(), equalTo(1));
+                }
+                nodeToDisconnect.heal();
+            });
+        }
+    }
+
+    public void testBeginPollingClusterFormationInfoCancel() {
+        /*
+         * This test sets up a 4-node cluster (3 master eligible). We call beginPollingClusterFormationInfo() on each node. We then
+         * cancel all tasks. This simulates what will happen most often in practice -- polling is triggered when the master node goes
+         * null, and then polling is cancelled when a new master node is elected within 10 seconds. We then simulate the cluster running
```

Comment:
"polling is cancelled when a new master node is elected within 10 seconds"

I think polling is cancelled immediately after a master node is detected. Maybe I'm misunderstanding what you meant here by the 10 seconds delay?